File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/FlutterMapView.java
package com.valentingrigorean.arcgis_maps_flutter.map;

import android.content.Context;
import android.graphics.Bitmap;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.Nullable;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.geometry.Geometry;
import com.esri.arcgisruntime.geometry.Point;
import com.esri.arcgisruntime.mapping.ArcGISMap;
import com.esri.arcgisruntime.mapping.SelectionProperties;
import com.esri.arcgisruntime.mapping.TimeExtent;
import com.esri.arcgisruntime.mapping.Viewpoint;
import com.esri.arcgisruntime.mapping.view.DrawStatusChangedListener;
import com.esri.arcgisruntime.mapping.view.GraphicsOverlay;
import com.esri.arcgisruntime.mapping.view.LayerViewStateChangedListener;
import com.esri.arcgisruntime.mapping.view.LocationDisplay;
import com.esri.arcgisruntime.mapping.view.MapScaleChangedListener;
import com.esri.arcgisruntime.mapping.view.MapView;
import com.esri.arcgisruntime.mapping.view.TimeExtentChangedListener;
import com.esri.arcgisruntime.mapping.view.ViewpointChangedListener;
import com.esri.arcgisruntime.util.ListenableList;

import java.util.ArrayList;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class FlutterMapView extends FrameLayout implements FlutterMapViewDelegate {
    private final MapView mapView;
    private final ArrayList<ViewpointChangedListener> viewpointChangedListeners = new ArrayList<>();
    private final ArrayList<DrawStatusChangedListener> drawStatusChangedListeners = new ArrayList<>();
    private final ArrayList<LayerViewStateChangedListener> layerViewStateChangedListeners = new ArrayList<>();
    private final ArrayList<MapScaleChangedListener> mapScaleChangedListeners = new ArrayList<>();
    private final ArrayList<TimeExtentChangedListener> timeExtentChangedListeners = new ArrayList<>();

    private boolean isDisposed = false;

    public FlutterMapView(Context context, AttributeSet attrs) {
        super(context, attrs);
        mapView = new MapView(context);
        setupMapView();
    }

    public FlutterMapView(Context context) {
        super(context);
        mapView = new MapView(context);
        setupMapView();
    }


    @Override
    protected void finalize() throws Throwable {
        dispose();
        super.finalize();
    }

    @Override
    public void dispose() {
        if (isDisposed) {
            return;
        }
        isDisposed = true;
        for (final ViewpointChangedListener listener : viewpointChangedListeners) {
            mapView.removeViewpointChangedListener(listener);
        }
        viewpointChangedListeners.clear();
        for (final DrawStatusChangedListener listener : drawStatusChangedListeners) {
            mapView.removeDrawStatusChangedListener(listener);
        }
        drawStatusChangedListeners.clear();
        for (final LayerViewStateChangedListener listener : layerViewStateChangedListeners) {
            mapView.removeLayerViewStateChangedListener(listener);
        }
        layerViewStateChangedListeners.clear();
        for (final MapScaleChangedListener listener : mapScaleChangedListeners) {
            mapView.removeMapScaleChangedListener(listener);
        }
        mapScaleChangedListeners.clear();
        for (final TimeExtentChangedListener listener : timeExtentChangedListeners) {
            mapView.removeTimeExtentChangedListener(listener);
        }
        timeExtentChangedListeners.clear();


        mapView.dispose();
    }

    @Override
    public void resume() {
        if (isDisposed) {
            return;
        }
        mapView.resume();
    }

    @Override
    public void pause() {
        if (isDisposed) {
            return;
        }
        mapView.pause();
    }

    @Nullable
    @Override
    public MapView getMapView() {
        return isDisposed ? null : mapView;
    }

    @Override
    public SelectionProperties getSelectionProperties() {
        if (isDisposed) {
            return null;
        }
        return mapView.getSelectionProperties();
    }

    @Override
    public ArcGISMap getMap() {
        return isDisposed ? null : mapView.getMap();
    }

    @Override
    public void setMap(ArcGISMap map) {
        if (isDisposed) {
            return;
        }
        mapView.setMap(map);
    }

    @Override
    public MapView.InteractionOptions getInteractionOptions() {
        return isDisposed ? null : mapView.getInteractionOptions();
    }

    @Override
    public void setAttributionTextVisible(boolean visible) {
        if (isDisposed) {
            return;
        }
        mapView.setAttributionTextVisible(visible);
    }

    @Override
    public void setViewInsets(double left, double top, double right, double bottom) {
        if (isDisposed) {
            return;
        }
        mapView.setViewInsets(left, top, right, bottom);
    }

    @Override
    public double getMapScale() {
        if (isDisposed) {
            return 0;
        }
        try {
            return mapView.getMapScale();
        } catch (Exception e) {
            Log.e("FlutterMapView", "getMapScale", e);
            return 0;
        }
    }

    @Override
    public double getMapRotation() {
        if (isDisposed) {
            return 0;
        }
        try {
            return mapView.getMapRotation();
        } catch (Exception e) {
            Log.e("FlutterMapView", "getMapRotation", e);
            return 0;
        }
    }

    @Override
    public LocationDisplay getLocationDisplay() {
        return isDisposed ? null : mapView.getLocationDisplay();
    }

    @Override
    public ListenableFuture<Bitmap> exportImageAsync() {
        return isDisposed ? new ValueListenableFutureExt<>(null) : mapView.exportImageAsync();
    }

    @Override
    public ListenableList<GraphicsOverlay> getGraphicsOverlays() {
        return isDisposed ? null : mapView.getGraphicsOverlays();
    }

    @Override
    public Viewpoint getCurrentViewpoint(Viewpoint.Type type) {
        return isDisposed ? null : mapView.getCurrentViewpoint(type);
    }

    @Override
    public ListenableFuture<Boolean> setViewpointAsync(Viewpoint viewpoint) {
        return isDisposed ? new ValueListenableFutureExt<>(false) : mapView.setViewpointAsync(viewpoint);
    }

    @Override
    public ListenableFuture<Boolean> setViewpointAsync(Viewpoint viewpoint, float durationSeconds) {
        return isDisposed ? new ValueListenableFutureExt<>(false) : mapView.setViewpointAsync(viewpoint, durationSeconds);
    }

    @Override
    public ListenableFuture<Boolean> setViewpointGeometryAsync(Geometry geometry) {
        return isDisposed ? new ValueListenableFutureExt<>(false) : mapView.setViewpointGeometryAsync(geometry);
    }

    @Override
    public ListenableFuture<Boolean> setViewpointGeometryAsync(Geometry geometry, double padding) {
        return isDisposed ? new ValueListenableFutureExt<>(false) : mapView.setViewpointGeometryAsync(geometry, padding);
    }

    @Override
    public ListenableFuture<Boolean> setViewpointCenterAsync(Point point, double scale) {
        return isDisposed ? new ValueListenableFutureExt<>(false) : mapView.setViewpointCenterAsync(point, scale);
    }

    @Override
    public ListenableFuture<Boolean> setViewpointScaleAsync(double scale) {
        return isDisposed ? new ValueListenableFutureExt<>(false) : mapView.setViewpointScaleAsync(scale);
    }

    @Override
    public ListenableFuture<Boolean> setViewpointRotationAsync(double rotation) {
        return isDisposed ? new ValueListenableFutureExt<>(false) : mapView.setViewpointRotationAsync(rotation);
    }

    @Override
    public android.graphics.Point locationToScreen(Point point) {
        return isDisposed ? null : mapView.locationToScreen(point);
    }

    @Override
    public Point screenToLocation(android.graphics.Point point) {
        return isDisposed ? null : mapView.screenToLocation(point);
    }

    @Override
    public TimeExtent getTimeExtent() {
        return isDisposed ? null : mapView.getTimeExtent();
    }

    @Override
    public void setTimeExtent(TimeExtent timeExtent) {
        if (isDisposed) {
            return;
        }
        mapView.setTimeExtent(timeExtent);
    }

    @Override
    public void addViewpointChangedListener(ViewpointChangedListener listener) {
        if (isDisposed) {
            return;
        }
        viewpointChangedListeners.add(listener);
        mapView.addViewpointChangedListener(listener);
    }

    @Override
    public void removeViewpointChangedListener(ViewpointChangedListener listener) {
        if (isDisposed) {
            return;
        }
        viewpointChangedListeners.remove(listener);
        mapView.removeViewpointChangedListener(listener);
    }

    @Override
    public void addDrawStatusChangedListener(DrawStatusChangedListener listener) {
        if (isDisposed) {
            return;
        }
        drawStatusChangedListeners.add(listener);
        mapView.addDrawStatusChangedListener(listener);
    }

    @Override
    public void removeDrawStatusChangedListener(DrawStatusChangedListener listener) {
        if (isDisposed) {
            return;
        }
        drawStatusChangedListeners.remove(listener);
        mapView.removeDrawStatusChangedListener(listener);
    }

    @Override
    public void addLayerViewStateChangedListener(LayerViewStateChangedListener listener) {
        if (isDisposed) {
            return;
        }
        layerViewStateChangedListeners.add(listener);
        mapView.addLayerViewStateChangedListener(listener);
    }

    @Override
    public void removeLayerViewStateChangedListener(LayerViewStateChangedListener listener) {
        if (isDisposed) {
            return;
        }
        layerViewStateChangedListeners.remove(listener);
        mapView.removeLayerViewStateChangedListener(listener);
    }

    @Override
    public void addMapScaleChangedListener(MapScaleChangedListener listener) {
        if (isDisposed) {
            return;
        }
        mapScaleChangedListeners.add(listener);
        mapView.addMapScaleChangedListener(listener);
    }

    @Override
    public void removeMapScaleChangedListener(MapScaleChangedListener listener) {
        if (isDisposed) {
            return;
        }
        mapScaleChangedListeners.remove(listener);
        mapView.removeMapScaleChangedListener(listener);
    }

    @Override
    public void addTimeExtentChangedListener(TimeExtentChangedListener listener) {
        if (isDisposed) {
            return;
        }
        timeExtentChangedListeners.add(listener);
        mapView.addTimeExtentChangedListener(listener);
    }

    @Override
    public void removeTimeExtentChangedListener(TimeExtentChangedListener listener) {
        if (isDisposed) {
            return;
        }
        timeExtentChangedListeners.remove(listener);
        mapView.removeTimeExtentChangedListener(listener);
    }

    private void setupMapView() {
        addView(mapView, new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT));
    }


    private class ValueListenableFutureExt<T> implements ListenableFuture<T> {
        private final T value;

        private ValueListenableFutureExt(T value) {
            this.value = value;
        }

        @Override
        public void addDoneListener(Runnable runnable) {
            runnable.run();
        }

        @Override
        public boolean removeDoneListener(Runnable runnable) {
            return false;
        }

        @Override
        public boolean cancel(boolean b) {
            return true;
        }

        @Override
        public boolean isCancelled() {
            return false;
        }

        @Override
        public boolean isDone() {
            return true;
        }

        @Override
        public T get() throws ExecutionException, InterruptedException {
            return value;
        }

        @Override
        public T get(long l, TimeUnit timeUnit) throws ExecutionException, InterruptedException, TimeoutException {
            return value;
        }
    }
}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/FlutterMapView.kt
package com.valentingrigorean.arcgis_maps_flutter.map

import android.content.Context
import android.graphics.Bitmap
import android.util.AttributeSet
import android.util.Log
import android.view.ViewGroup
import android.widget.FrameLayout
import com.esri.arcgisruntime.concurrent.ListenableFuture
import com.esri.arcgisruntime.geometry.Geometry
import com.esri.arcgisruntime.geometry.Point
import com.esri.arcgisruntime.mapping.ArcGISMap
import com.esri.arcgisruntime.mapping.SelectionProperties
import com.esri.arcgisruntime.mapping.TimeExtent
import com.esri.arcgisruntime.mapping.Viewpoint
import com.esri.arcgisruntime.mapping.view.DrawStatusChangedListener
import com.esri.arcgisruntime.mapping.view.GraphicsOverlay
import com.esri.arcgisruntime.mapping.view.LayerViewStateChangedListener
import com.esri.arcgisruntime.mapping.view.LocationDisplay
import com.esri.arcgisruntime.mapping.view.MapScaleChangedListener
import com.esri.arcgisruntime.mapping.view.MapView
import com.esri.arcgisruntime.mapping.view.TimeExtentChangedListener
import com.esri.arcgisruntime.mapping.view.ViewpointChangedListener
import com.esri.arcgisruntime.util.ListenableList
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

class FlutterMapView : FrameLayout, FlutterMapViewDelegate {
    private override val mapView: MapView
    private val viewpointChangedListeners = ArrayList<ViewpointChangedListener>()
    private val drawStatusChangedListeners = ArrayList<DrawStatusChangedListener>()
    private val layerViewStateChangedListeners = ArrayList<LayerViewStateChangedListener>()
    private val mapScaleChangedListeners = ArrayList<MapScaleChangedListener>()
    private val timeExtentChangedListeners = ArrayList<TimeExtentChangedListener>()
    private var isDisposed = false

    constructor(context: Context?, attrs: AttributeSet?) : super(context!!, attrs) {
        mapView = MapView(context)
        setupMapView()
    }

    constructor(context: Context?) : super(context!!) {
        mapView = MapView(context)
        setupMapView()
    }

    @Throws(Throwable::class)
    protected fun finalize() {
        dispose()
        super.finalize()
    }

    override fun dispose() {
        if (isDisposed) {
            return
        }
        isDisposed = true
        for (listener in viewpointChangedListeners) {
            mapView.removeViewpointChangedListener(listener)
        }
        viewpointChangedListeners.clear()
        for (listener in drawStatusChangedListeners) {
            mapView.removeDrawStatusChangedListener(listener)
        }
        drawStatusChangedListeners.clear()
        for (listener in layerViewStateChangedListeners) {
            mapView.removeLayerViewStateChangedListener(listener)
        }
        layerViewStateChangedListeners.clear()
        for (listener in mapScaleChangedListeners) {
            mapView.removeMapScaleChangedListener(listener)
        }
        mapScaleChangedListeners.clear()
        for (listener in timeExtentChangedListeners) {
            mapView.removeTimeExtentChangedListener(listener)
        }
        timeExtentChangedListeners.clear()
        mapView.dispose()
    }

    override fun resume() {
        if (isDisposed) {
            return
        }
        mapView.resume()
    }

    override fun pause() {
        if (isDisposed) {
            return
        }
        mapView.pause()
    }

    override fun getMapView(): MapView? {
        return if (isDisposed) null else mapView
    }

    override val selectionProperties: SelectionProperties?
        get() = if (isDisposed) {
            null
        } else mapView.selectionProperties
    override var map: ArcGISMap?
        get() = if (isDisposed) null else mapView.map
        set(map) {
            if (isDisposed) {
                return
            }
            mapView.map = map
        }
    override val interactionOptions: MapView.InteractionOptions?
        get() = if (isDisposed) null else mapView.interactionOptions

    override fun setAttributionTextVisible(visible: Boolean) {
        if (isDisposed) {
            return
        }
        mapView.isAttributionTextVisible = visible
    }

    override fun setViewInsets(left: Double, top: Double, right: Double, bottom: Double) {
        if (isDisposed) {
            return
        }
        mapView.setViewInsets(left, top, right, bottom)
    }

    override val mapScale: Double
        get() = if (isDisposed) {
            0
        } else try {
            mapView.mapScale
        } catch (e: Exception) {
            Log.e("FlutterMapView", "getMapScale", e)
            0
        }
    override val mapRotation: Double
        get() = if (isDisposed) {
            0
        } else try {
            mapView.mapRotation
        } catch (e: Exception) {
            Log.e("FlutterMapView", "getMapRotation", e)
            0
        }
    override val locationDisplay: LocationDisplay?
        get() = if (isDisposed) null else mapView.locationDisplay

    override fun exportImageAsync(): ListenableFuture<Bitmap?> {
        return if (isDisposed) ValueListenableFutureExt(null) else mapView.exportImageAsync()
    }

    override val graphicsOverlays: ListenableList<GraphicsOverlay?>?
        get() = if (isDisposed) null else mapView.graphicsOverlays

    override fun getCurrentViewpoint(type: Viewpoint.Type?): Viewpoint? {
        return if (isDisposed) null else mapView.getCurrentViewpoint(type)
    }

    override fun setViewpointAsync(viewpoint: Viewpoint?): ListenableFuture<Boolean?> {
        return if (isDisposed) ValueListenableFutureExt<Boolean>(false) else mapView.setViewpointAsync(
            viewpoint
        )
    }

    override fun setViewpointAsync(
        viewpoint: Viewpoint?,
        durationSeconds: Float
    ): ListenableFuture<Boolean?> {
        return if (isDisposed) ValueListenableFutureExt<Boolean>(false) else mapView.setViewpointAsync(
            viewpoint,
            durationSeconds
        )
    }

    override fun setViewpointGeometryAsync(geometry: Geometry?): ListenableFuture<Boolean?> {
        return if (isDisposed) ValueListenableFutureExt<Boolean>(false) else mapView.setViewpointGeometryAsync(
            geometry
        )
    }

    override fun setViewpointGeometryAsync(
        geometry: Geometry?,
        padding: Double
    ): ListenableFuture<Boolean?> {
        return if (isDisposed) ValueListenableFutureExt<Boolean>(false) else mapView.setViewpointGeometryAsync(
            geometry,
            padding
        )
    }

    override fun setViewpointCenterAsync(point: Point?, scale: Double): ListenableFuture<Boolean?> {
        return if (isDisposed) ValueListenableFutureExt<Boolean>(false) else mapView.setViewpointCenterAsync(
            point,
            scale
        )
    }

    override fun setViewpointScaleAsync(scale: Double): ListenableFuture<Boolean?> {
        return if (isDisposed) ValueListenableFutureExt<Boolean>(false) else mapView.setViewpointScaleAsync(
            scale
        )
    }

    override fun setViewpointRotationAsync(rotation: Double): ListenableFuture<Boolean?> {
        return if (isDisposed) ValueListenableFutureExt<Boolean>(false) else mapView.setViewpointRotationAsync(
            rotation
        )
    }

    override fun locationToScreen(point: Point?): android.graphics.Point? {
        return if (isDisposed) null else mapView.locationToScreen(point)
    }

    override fun screenToLocation(point: android.graphics.Point?): Point? {
        return if (isDisposed) null else mapView.screenToLocation(point)
    }

    override var timeExtent: TimeExtent?
        get() = if (isDisposed) null else mapView.timeExtent
        set(timeExtent) {
            if (isDisposed) {
                return
            }
            mapView.timeExtent = timeExtent
        }

    override fun addViewpointChangedListener(listener: ViewpointChangedListener) {
        if (isDisposed) {
            return
        }
        viewpointChangedListeners.add(listener)
        mapView.addViewpointChangedListener(listener)
    }

    override fun removeViewpointChangedListener(listener: ViewpointChangedListener) {
        if (isDisposed) {
            return
        }
        viewpointChangedListeners.remove(listener)
        mapView.removeViewpointChangedListener(listener)
    }

    override fun addDrawStatusChangedListener(listener: DrawStatusChangedListener) {
        if (isDisposed) {
            return
        }
        drawStatusChangedListeners.add(listener)
        mapView.addDrawStatusChangedListener(listener)
    }

    override fun removeDrawStatusChangedListener(listener: DrawStatusChangedListener) {
        if (isDisposed) {
            return
        }
        drawStatusChangedListeners.remove(listener)
        mapView.removeDrawStatusChangedListener(listener)
    }

    override fun addLayerViewStateChangedListener(listener: LayerViewStateChangedListener) {
        if (isDisposed) {
            return
        }
        layerViewStateChangedListeners.add(listener)
        mapView.addLayerViewStateChangedListener(listener)
    }

    override fun removeLayerViewStateChangedListener(listener: LayerViewStateChangedListener) {
        if (isDisposed) {
            return
        }
        layerViewStateChangedListeners.remove(listener)
        mapView.removeLayerViewStateChangedListener(listener)
    }

    override fun addMapScaleChangedListener(listener: MapScaleChangedListener) {
        if (isDisposed) {
            return
        }
        mapScaleChangedListeners.add(listener)
        mapView.addMapScaleChangedListener(listener)
    }

    override fun removeMapScaleChangedListener(listener: MapScaleChangedListener) {
        if (isDisposed) {
            return
        }
        mapScaleChangedListeners.remove(listener)
        mapView.removeMapScaleChangedListener(listener)
    }

    override fun addTimeExtentChangedListener(listener: TimeExtentChangedListener) {
        if (isDisposed) {
            return
        }
        timeExtentChangedListeners.add(listener)
        mapView.addTimeExtentChangedListener(listener)
    }

    override fun removeTimeExtentChangedListener(listener: TimeExtentChangedListener) {
        if (isDisposed) {
            return
        }
        timeExtentChangedListeners.remove(listener)
        mapView.removeTimeExtentChangedListener(listener)
    }

    private fun setupMapView() {
        addView(
            mapView, LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
            )
        )
    }

    private inner class ValueListenableFutureExt<T>(private val value: T) : ListenableFuture<T> {
        override fun addDoneListener(runnable: Runnable) {
            runnable.run()
        }

        override fun removeDoneListener(runnable: Runnable): Boolean {
            return false
        }

        override fun cancel(b: Boolean): Boolean {
            return true
        }

        override fun isCancelled(): Boolean {
            return false
        }

        override fun isDone(): Boolean {
            return true
        }

        @Throws(ExecutionException::class, InterruptedException::class)
        override fun get(): T {
            return value
        }

        @Throws(ExecutionException::class, InterruptedException::class, TimeoutException::class)
        override fun get(l: Long, timeUnit: TimeUnit): T {
            return value
        }
    }
}